stl/inc/ranges

```diff
@@ -7656,6 +7662,412 @@ namespace ranges {
     } // namespace views
 
+    _EXPORT_STD template <move_constructible _Func, input_range... _ViewTypes>
+        requires ((view<_ViewTypes> && ...) && (sizeof...(_ViewTypes) > 0)
+                  && is_object_v<_Func> && regular_invocable<_Func&, range_reference_t<_ViewTypes>...>
+                  && _Can_reference<invoke_result_t<_Func&, range_reference_t<_ViewTypes>...>>)
+    class zip_transform_view : public view_interface<zip_transform_view<_Func, _ViewTypes...>> {
+    private:
+        using _Inner_view = zip_view<_ViewTypes...>;
+
+        /* [[no_unique_address]] */ _Movable_box<_Func> _Function;
+        /* [[no_unique_address]] */ _Inner_view _Zip;
+
+        template <bool _IsConst>
+        using _Ziperator = iterator_t<_Maybe_const<_IsConst, _Inner_view>>;
+
+        template <bool _IsConst>
+        using _Zentinel = sentinel_t<_Maybe_const<_IsConst, _Inner_view>>;
+
+        template <bool _IsConst>
+        struct _Category_base {};
+
+        template <bool _IsConst>
+            requires forward_range<_Maybe_const<_IsConst, _Inner_view>>
+        struct _Category_base<_IsConst> {
+        private:
+            enum class _Category_id {
+                _Random_access_iterator,
+                _Bidirectional_iterator,
+                _Forward_iterator,
+                _Input_iterator
+            };
+
+            static constexpr _Category_id _Curr_id = []() noexcept {
+                using _Invoke_t = invoke_result_t<_Maybe_const<_IsConst, _Func>&,
+                    range_reference_t<_Maybe_const<_IsConst, _ViewTypes>>...>;
+
+                if constexpr (!is_reference_v<_Invoke_t>) {
+                    return _Category_id::_Input_iterator;
+                } else {
+                    constexpr auto _Check_tag_type_closure = []<typename _Tag_type>() noexcept {
+                        // clang-format off
+                        return (derived_from<typename
+                            iterator_traits<iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>::iterator_category,
+                            _Tag_type>
+                        && ...);
+                        // clang-format on
+                    };
+
+                    if constexpr (_Check_tag_type_closure.template operator()<random_access_iterator_tag>()) {
+                        return _Category_id::_Random_access_iterator;
+                    } else if constexpr (_Check_tag_type_closure.template operator()<bidirectional_iterator_tag>()) {
+                        return _Category_id::_Bidirectional_iterator;
+                    } else if constexpr (_Check_tag_type_closure.template operator()<forward_iterator_tag>()) {
+                        return _Category_id::_Forward_iterator;
+                    } else {
+                        return _Category_id::_Input_iterator;
+                    }
+                }
+            }();
+
+        public:
+            using iterator_category =

```

Comment:
Can we use `using iterator_category = decltype([]{ ... }());`and make the lambda expression return `input_iterator_tag{}` etc.?

Reply:
That's a great idea, actually! I never even considered that. I'll be sure to add `decltype([]{ ... }())` to my long list of C++ idioms...